janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java

```diff
@@ -0,0 +1,243 @@
+    }
+
+    protected int[] getOffsetIndex() {
+        return offsetIndex;
+    }
+
+    protected byte[] getRawData() {
+        return rawData;
+    }
+
+    protected void setRawData(final byte[] rawData) {
+        this.rawData = rawData;
+    }
+
+    protected void setOffsetIndex(final int[] offsetIndex) {
+        this.offsetIndex = offsetIndex;
+    }
+
+    public boolean isEmpty() {
+        return offsetIndex.length == 0;
+    }
+
+    public int getIndex(final StaticBuffer column) {
+        //binary search on column names
+        int low = 0;
+        int high = offsetIndex.length - 1;
+        int compare;
+        int mid;
+
+        while (low <= high) {
+            mid = (low + high) >>> 1;
+
+            //NOTE: we used to do getColumn(mid).compareTo(column), where compareTo internally asserts that column is a StaticARRAYBuffer, so
+            // apparently we don't have to worry about column name being in anything but array-backed buffer
+            //So we can compare relevant array parts directly, avoiding the churn of new StaticArrayBuffers created by getColumn()
+            Preconditions.checkArgument(column instanceof StaticArrayBuffer);
+            compare = -((StaticArrayBuffer) column).compareTo(rawData, offsetIndex[mid] + 1, offsetIndex[mid] + 1 + rawData[offsetIndex[mid]]);
+
+            if (compare < 0) {
+                low = mid + 1;
+            } else if (compare > 0) {
+                high = mid - 1;
+            } else {
+                return mid; // key found
+            }
+        }
+        return -(low + 1);  // key not found.
+    }
+
+    private int getEntryLength(final int index) {
+        if (index < offsetIndex.length - 1) {
+            return offsetIndex[index + 1] - offsetIndex[index];
+        } else {
+            return rawData.length - offsetIndex[index];
+        }
+    }
+
+    private int getEntryEndOffset(final int index) {
+        if (index < offsetIndex.length - 1) {
+            return offsetIndex[index + 1];
+        } else {
+            return rawData.length;
+        }
+    }
+
+    public Entry get(final int index) {
+        final int entryBufLen = getEntryLength(index);
+
+        final ByteBuffer entryBuffer = ByteBuffer.wrap(rawData, offsetIndex[index], entryBufLen);
+
+        final byte valPos = entryBuffer.get();
+
+        final byte[] col = new byte[valPos];
+        entryBuffer.get(col);
+
+        final byte[] val = new byte[entryBufLen - valPos - 1];
+        entryBuffer.get(val);
+
+        return StaticArrayEntry.of(StaticArrayBuffer.of(col), StaticArrayBuffer.of(val));
+    }
+
+    public Entry getNoCopy(final int index) {
+        return new StaticArrayEntry(rawData, offsetIndex[index] + 1, getEntryEndOffset(index), rawData[offsetIndex[index]]);
+    }
+
+    public int numEntries() {
+        return offsetIndex.length;
+    }
+
+    public int byteSize() {
+        return rawData.length + offsetIndex.length * Integer.BYTES + 16;
```

Comment:
(question) 
Do I understand it right that when we execute `byteSize` in `MultiPageEntryBuffer` it should return the sum of all buffers, where each bugger takes:
`offsetIndex.length * Integer.BYTES + 32 + rawData.length`? Is it a right formula?
If yes, then why do you add +16 here and +16 in `byteSize` method of `MultiPageEntryBuffer`? Why not to add +32 just here?

Reply:
Usually most of the buffers in a typical Janusgraph store will be less than 500 entries (so one page), and thus using "multi-page" format for all of them would be a waste, since it requires more overhead bytes to store the ArrayList of pages, etc. So we start with SinglePageEntryBuffer, and only convert to MultiPage if we have to.

So for SinglePageEntryBuffer the total byte size will be exactly the same as for a single page, but for multi-page buffer the total byte size is (roughly) the sum of each page's size plus the size of the reference to the page from the list of pages (plus the overhead of the ArrayList internals).

Reply:
Thank you for the explanation @dk-github !